turbopack/crates/turbo-tasks-fs/src/glob.rs

```diff
@@ -39,390 +13,132 @@ enum GlobPart {
 // - {a/**,*}/file = Alternatives([File(a), PathSeparator, AnyDirectories], [AnyFile]),
 //   PathSeparator, File(file)
 
-// Note: a/**/b does match a/b, so we need some special logic about path
-// separators
-
-#[turbo_tasks::value]
-#[derive(Debug, Clone)]
+#[turbo_tasks::value(eq = "manual")]
+#[serde(into = "GlobForm", try_from = "GlobForm")]
+#[derive(Clone, Debug)]
 pub struct Glob {
-    expression: Vec<GlobPart>,
+    // Store the raw glob strings to support equality and serialization
+    raw: Vec<RcStr>,
+    #[turbo_tasks(trace_ignore)]
+    implementation: GlobImpl,
+}
+#[derive(Clone, Debug)]
+enum GlobImpl {
+    Set(GlobSet),
+    Single(GlobMatcher),
```

Comment:
What's the advantage to having a separate codepath for `Single`? Is there much overhead to constructing a `GlobSet` with a single item in it?

Reply:
It is interesting, the strategies of the two implementations are actually quite different and afaict BurntSushi decided the 'set' strategy wasn't worthwhile for single matches: https://github.com/BurntSushi/ripgrep/blob/6dfaec03e830892e787686917509c17860456db1/crates/globset/src/glob.rs#L269. which is why there are 2 separate types.  tl;dr; the singleton strategy is actually faster.

so i figured we would just maintain that specialization.  The memory overhead of a single glob is _slightly_ smaller as well.